File: main_project/tripod-new-branch/src/cpp/configclient/CacheManagerConfigClient.cpp
#include <boost/algorithm/string/split.hpp>
#include <boost/algorithm/string/classification.hpp>
#include <Common.h>
#include <ConfigData.h>
#include <CacheManagerConfigClient.h>

namespace com {
namespace renren {
namespace tripod {

CacheManagerConfigClient::CacheManagerConfigClient(const std::string& zkAddress) : ConfigClient(zkAddress, TRIPOD_MANAGER) {
    MCE_INFO("CacheManagerConfigClient::CacheManagerConfigClient zkAddress " << zkAddress);
    ic_ = Ice::initialize();
}

bool CacheManagerConfigClient::update(const ConfigDataPtr& newConfig) {
    MCE_INFO("CacheManagerConfigClient::update");
    std::map<std::string, ConfigDataPtr>::const_iterator findIt = newConfig->childrenDatas_.find(TRIPOD_MANAGER);
    if(findIt == newConfig->childrenDatas_.end()) {
        MCE_WARN("CacheManagerConfigClient::update no cachemanager in config!");
        return false;
    }
    std::map<std::string, CacheManagerPrx> prxs;
    for(std::map<std::string, ConfigDataPtr>::const_iterator it = findIt->second->childrenDatas_.begin();
            it != findIt->second->childrenDatas_.end(); ++it) {
            
        std::vector<std::string> parts;
        boost::algorithm::split(parts, it->second->name_, boost::algorithm::is_any_of(":"));
        if(parts.size() != 2) {
            std::ostringstream oss;
            oss << "CacheManagerConfigClient::update Exception the config string" << it->second->name_
                << " not match the format: <host>:<port>";
            MCE_WARN(oss.str());
            return false;
        }
        std::string address = parts[0];
        std::string port = parts[1];
        std::ostringstream os;
        os<<"M:tcp -h "<<address<<" -p "<<port<<" -t 1000";
        prxs.insert(make_pair(it->second->name_, CacheManagerPrx::uncheckedCast(ic_->stringToProxy(os.str()))));
    }

    hashHolder_.renew(prxs);

    return true;
}

std::map<CacheManagerPrx, KeySeq> CacheManagerConfigClient::getPrxToKeysMap(const KeySeq& keys) {
    return hashHolder_.get(keys);
}

std::map<std::string, CacheManagerPrx> CacheManagerConfigClient::getOriginalPrxMap() {
    return hashHolder_.getOriginalPrxMap();
}

}
}
}

File: main_project/tripod-new-branch/src/java/TripodConfig/src/main/java/com/renren/tripod/configclient/CacheManagerConfigClient.java
package com.renren.tripod.configclient;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import com.renren.tripod.CacheManagerPrx;
import com.renren.tripod.CacheManagerPrxHelper;
import com.renren.tripod.configutil.ConfigClient;
import com.renren.tripod.configutil.ConfigData;

public class CacheManagerConfigClient extends ConfigClient {
	private static org.apache.commons.logging.Log _logger = org.apache.commons.logging.LogFactory
			.getLog(CacheManagerConfigClient.class);

	private Ice.Communicator _ic = Ice.Util.initialize();
	private HashHolder<CacheManagerPrx> _hashHolder = new HashHolder<CacheManagerPrx>(
			100);

	public CacheManagerConfigClient(String zkAddress, String identifier) {
		super(zkAddress, identifier);
		_logger.info("CacheManagerConfigClient.CacheManagerConfigClient zkAddress "
				+ zkAddress + " identifier " + identifier);
	}

	@Override
	protected boolean update(ConfigData newConfig) {
		_logger.info("CacheManagerConfigClient.update");
		ConfigData cmConfig = newConfig._childrenDatas
				.get(ConfigData.TRIPOD_MANAGER);
		if (cmConfig == null) {
			_logger.error("CacheManagerConfigClient.update cmConfig == null!");
			return false;
		}
		Map<String, CacheManagerPrx> clients = new HashMap<String, CacheManagerPrx>();
		for (Entry<String, ConfigData> ent : cmConfig._childrenDatas.entrySet()) {
			String[] parts = ent.getKey().split(":");
			if (parts.length != 2) {
				return false;
			}
			CacheManagerPrx prx = CacheManagerPrxHelper.uncheckedCast(_ic
					.stringToProxy("M:tcp -h " + parts[0] + " -p " + parts[1] + " -t " + 1000));
			if (prx == null) {
				_logger.error("CacheManagerConfigClient.update prx == null for string "
						+ ent.getKey());
				return false;
			}
			clients.put(ent.getKey(), prx);
		}
		_hashHolder.renew(clients);
		return true;
	}

	public CacheManagerPrx getPrx(String key) {
		return _hashHolder.get(key);
	}

	public Map<CacheManagerPrx, List<String>> getPrxToKeysMap(List<String> keys) {
		return _hashHolder.get(keys);
	}

}
